### PR TITLE
[FEATURE] Experimental filesystem `DataAsset`s `path` in batch request options and batch metadata

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-1679
- Add `path` to batch request options and batch metadata for batch requests with files as batches
- Allow filesystem-type assets to be added without passing a `regex`
- Handle `regex` and `order_by` type conversion outside of `pydantic` model

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/core/util.py
Comment (NathanFarmer) [inline great_expectations/core/util.py:335]: All `pathlib` classes inherit from `PurePath` per the diagram at the top of this page: https://docs.python.org/3/library/pathlib.html

File: great_expectations/experimental/datasources/filesystem_data_asset.py
Comment (NathanFarmer) [inline great_expectations/experimental/datasources/filesystem_data_asset.py:58]: This change enables the case where our `regex` references a directory structure, as opposed to a filename.

Comment (NathanFarmer) [reply great_expectations/experimental/datasources/filesystem_data_asset.py:58]: An integration test was added to ensure this works.

File: great_expectations/experimental/datasources/file_path_data_asset.py
Comment (NathanFarmer) [inline great_expectations/experimental/datasources/file_path_data_asset.py:108]: In looking for a way to get type hints at the API level for `regex` and `order_by` I learned that the creators of `pydantic` [recommend that you do not perform type conversion in validators](https://github.com/pydantic/pydantic/discussions/3997), but instead do the type conversion outside of the `BaseModel`. This lead me to move the logic for `parse_regex_string` into a static method that can be leveraged by the `Datasource` `add_*` methods. We do not need to keep this logic in the validator. The side effect of removing it would just be that you have to pass the actual type if you don't want to use `type: ignore`.

File: tests/experimental/datasources/test_pandas_datasource.py
Comment (NathanFarmer) [inline tests/experimental/datasources/test_pandas_datasource.py]: This behavior was incorrect.

**Comment by alexsherstinsky** [comment]
@NathanFarmer Unfortunately, the merge conflicts have to be resolved a-priori...  For example, the testing of connection logic has changed.  In addition, `FilesystemDataAsset` will go away.  Maybe we can together update this branch after my next refactor, and then merge it?  These next steps will be coming very soon.  Happy to discuss.  Thank you.

File: great_expectations/experimental/datasources/file_path_data_asset.py
Comment (alexsherstinsky) [reply great_expectations/experimental/datasources/file_path_data_asset.py:108]: @Kilo59 Could you please review this reasoning?  To me, validator should only validate and not be utilized for "post init" type purposes; but I do not have enough experience with it in practice to be certain.  Thank you.

File: great_expectations/experimental/datasources/file_path_data_asset.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/file_path_data_asset.py:59]: @NathanFarmer I found a way to handle this using `batch_spec` (in #7197).  It would be great if we could merge this pull request after that one; otherwise, the merge can become very tedious (due to the refactoring).  Thank you.

File: great_expectations/experimental/datasources/filesystem_data_asset.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/filesystem_data_asset.py:45]: @NathanFarmer This module (`filesystem_data_asset.py`) is proposed to be removed in #7197 (its core functionality moved up to `FilePathDataConnector`); hence the belief that we should merge that PR first (assuming that there are no objections).  Thank you for your consideration.

File: great_expectations/experimental/datasources/filesystem_data_asset.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/filesystem_data_asset.py:156]: @NathanFarmer This code is proposed to be moved to `FilePathDataConnector` (it is general for all file-based assets).

File: great_expectations/experimental/datasources/filesystem_data_asset.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/filesystem_data_asset.py:203]: @NathanFarmer Instead of `base_directory`, we can use `PathBatchSpec` one level above (and thus remove this module).  This is proposed in #7197.

File: great_expectations/experimental/datasources/interfaces.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/interfaces.py:227]: @NathanFarmer FYI: I am going to attempt to move the job of sorting into `DataConnector` (not sure if this code will change.  Thanks.

File: great_expectations/experimental/datasources/pandas_filesystem_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/pandas_filesystem_datasource.py:5]: @NathanFarmer FYI: This module is changing in #7197 (quite a bit).

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/spark_datasource.py:14]: @NathanFarmer This module is being renamed in #7197 (to indicate that it is filesystem-only).  Thanks.

File: great_expectations/experimental/datasources/spark_datasource.py
Comment (alexsherstinsky) [inline great_expectations/experimental/datasources/spark_datasource.py:102]: @NathanFarmer This test can now be done through the `DataConnector` at the file-path level.

File: tests/experimental/datasources/test_spark_datasource.py
Comment (alexsherstinsky) [inline tests/experimental/datasources/test_spark_datasource.py:65]: @NathanFarmer This module is being renamed in #7197 (by the way, the method `test_add_csv_asset_to_datasource()` seems to already exist; is it being simply moved?).

**Review by alexsherstinsky — APPROVED** [review]
* Approving this, pending feedback from @Kilo59 on the question about the Pydantic validator.
* Also requesting if we could please let #7197 go through first (since it refactors quite a bit).

Thank you.

File: great_expectations/experimental/datasources/file_path_data_asset.py
Comment (NathanFarmer) [reply great_expectations/experimental/datasources/file_path_data_asset.py:59]: @alexsherstinsky I am not inclined to wait any longer to merge another one of your PRs.

**Pull request merged**
